# Data Science Regression Project: Predicting Home Prices in Banglore

Dataset is downloaded from here: https://www.kaggle.com/amitabhajoy/bengaluru-house-price-data

In [1]:
# Pandas is a dataframe 
import pandas as pd

In [2]:
# Numpy = number python, allows you to do matrix multiplication 
import numpy as np

In [3]:
#matplotlib = library to help make plots 
from matplotlib import pyplot as plt

In [4]:
%matplotlib inline

In [5]:
import matplotlib

In [6]:
matplotlib.rcParams["figure.figsize"]= (20,10)

## Data Load: Load banglore home prices into dataframe

In [7]:
# pd.read_csv - reads csv file
df1 = pd.read_csv("bengaluru_house_prices.csv")
# df1.head() - will show you the initial/top of the dataframe, avoid loading entire dataframe
df1.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [8]:
# Can set a number for head 
df1.head(6)

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   
5  Super built-up  Area  Ready To Move                Whitefield      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00  
5  DuenaTa       1170   2.0      1.0   38.00

In [9]:
# Can also use tail to see last/ending of dataframe
df1.tail()

area_type   availability               location       size  \
13315        Built-up  Area  Ready To Move             Whitefield  5 Bedroom   
13316  Super built-up  Area  Ready To Move          Richards Town      4 BHK   
13317        Built-up  Area  Ready To Move  Raja Rajeshwari Nagar      2 BHK   
13318  Super built-up  Area         18-Jun        Padmanabhanagar      4 BHK   
13319  Super built-up  Area  Ready To Move           Doddathoguru      1 BHK   

       society total_sqft  bath  balcony  price  
13315  ArsiaEx       3453   4.0      0.0  231.0  
13316      NaN       3600   5.0      NaN  400.0  
13317  Mahla T       1141   2.0      1.0   60.0  
13318  SollyCl       4689   4.0      1.0  488.0  
13319      NaN        550   1.0      1.0   17.0

In [10]:
df1.shape
# rows, columns

(13320, 9)

In [11]:
df1.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

In [12]:
# Show the unique variables in this column 
df1['area_type'].unique()

array(['Super built-up  Area', 'Plot  Area', 'Built-up  Area',
       'Carpet  Area'], dtype=object)

In [13]:
# Different way to count/show unique type 
df1.groupby('area_type')['area_type'].agg('count')

area_type
Built-up  Area          2418
Carpet  Area              87
Plot  Area              2025
Super built-up  Area    8790
Name: area_type, dtype: int64

In [14]:
# This will add up each unique category in the column 'area_type'
df1['area_type'].value_counts()

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64

In [15]:
df1['availability'].unique()

array(['19-Dec', 'Ready To Move', '18-May', '18-Feb', '18-Nov', '20-Dec',
       '17-Oct', '21-Dec', '19-Sep', '20-Sep', '18-Mar', '20-Feb',
       '18-Apr', '20-Aug', '18-Oct', '19-Mar', '17-Sep', '18-Dec',
       '17-Aug', '19-Apr', '18-Jun', '22-Dec', '22-Jan', '18-Aug',
       '19-Jan', '17-Jul', '18-Jul', '21-Jun', '20-May', '19-Aug',
       '18-Sep', '17-May', '17-Jun', '21-May', '18-Jan', '20-Mar',
       '17-Dec', '16-Mar', '19-Jun', '22-Jun', '19-Jul', '21-Feb',
       'Immediate Possession', '19-May', '17-Nov', '20-Oct', '20-Jun',
       '19-Feb', '21-Oct', '21-Jan', '17-Mar', '17-Apr', '22-May',
       '19-Oct', '21-Jul', '21-Nov', '21-Mar', '16-Dec', '22-Mar',
       '20-Jan', '21-Sep', '21-Aug', '14-Nov', '19-Nov', '15-Nov',
       '16-Jul', '15-Jun', '17-Feb', '20-Nov', '20-Jul', '16-Sep',
       '15-Oct', '15-Dec', '16-Oct', '22-Nov', '15-Aug', '17-Jan',
       '16-Nov', '20-Apr', '16-Jan', '14-Jul'], dtype=object)

In [16]:
df1['society'].unique()

array(['Coomee ', 'Theanmp', nan, ..., 'SJovest', 'ThhtsV ', 'RSntsAp'],
      dtype=object)

In [17]:
df1['balcony'].unique()

array([ 1.,  3., nan,  2.,  0.])

Out of the 9 columns, determine which ones are not needed and which ones are necessary 

Columns: 
1. area_type (text, ex: Super built-up Area, Plot Area)
2. availability (text, ex: Ready To Move,19-Dec) 
3. location (text, ex: Electronic City Phase II, Chikka Tirupathi)  
4. size (text/number, ex: 2 BHK, 4 Bedroom) 
5. society (text, ex: Coomee, Theanmp) 
6. total_sqft (number, ex: 1056, 2600) 
7. bath (number, ex: 2.0, 5.0) 
8. balcony (number, ex: 1.0, 3.0) 
9. price (number, ex: 39.07, 120.00) 

Of those, most likely the ones that matter will be

3. location - price varies dramatically on location (ex: beach, good location)
4. size - number of bedrooms is an important factor for a home
6. total_sqft - good indication on property size and worth 
7. bath - number of bathrooms affect housing price 
9. price - actual price helps predict the house worth depending on above variables 

The columns that do not matter too much
1. area_type - knowing the area type in this scenario does not seem to matter 
2. availability - knowning the date when its available to sell does not help 
5. society - societial views do not really affect pricing 
8. balcony - although this could affect pricing, not sure the correlation it could have 

In [18]:
df2 = df1.drop(['area_type', 'society', 'balcony', 'availability'], axis="columns")
df2.shape

(13320, 5)

In [19]:
df1.shape

(13320, 9)

# Dropping the columns that are not helpful, you can notice that the columns have reduced from 9 to 5 

df1.shape = (13320, 9)

df2.shape = (13320, 5)

Given this dataset, what are things we must do first in order to ensure the data is clean, organized, and usable 

1. remove any 0 or NaN/null values 
2. make sure each column values are uniform/similiar - sometimes values are in ranges
    - ex: 2100 - 2850, (9 Bedroom vs 4 BHK)
3. using business logic, remove any data that could be an outlier 

In [20]:
df2.isnull()

location   size  total_sqft   bath  price
0         False  False       False  False  False
1         False  False       False  False  False
2         False  False       False  False  False
3         False  False       False  False  False
4         False  False       False  False  False
...         ...    ...         ...    ...    ...
13315     False  False       False  False  False
13316     False  False       False  False  False
13317     False  False       False  False  False
13318     False  False       False  False  False
13319     False  False       False  False  False

[13320 rows x 5 columns]

In [21]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [22]:
df3 = df2.dropna()

In [23]:
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [24]:
df3.shape

(13246, 5)

df1.shape = (13320, 9)

df2.shape = (13320, 5)

df3.shape = (13246, 5) 

In [25]:
# create a new column... df3['new_column_name']  = .... 
# new column equals the value of another column given some additional logic 
# .apply (will apply this logic to each variable)
# lambda x is an anonymous function, can take any number of agruments but only 1 expression 
# lambda arguments: expression 
df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))
df3.bhk.unique()

C:\Users\15599\AppData\Local\Temp\ipykernel_13516\354817163.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df3['bhk'] = df3['size'].apply(lambda x: int(x.split(' ')[0]))


array([ 2,  4,  3,  6,  1,  8,  7,  5, 11,  9, 27, 10, 19, 16, 43, 14, 12,
       13, 18], dtype=int64)

Here you notice that the new value will be x.split(' ')[0]...

most of the value of "bhk" is "3 BHK" or "3 bedroom"

the x.split will split the variables by a space which in this case will be [3, bhk] or [3, bedroom] and then will take the first value [0] being the number

    -lmbda x: int(x.split(' ')[0]) 
    - x = "3 bhk"
    -int("3 bhk".split(' ')[0])
    -int(["3, bhk"][0])
    -int(["3"])
    -3

In [26]:
df3['bhk'].value_counts()

2     5527
3     4832
4     1395
1      649
5      353
6      221
7      100
8       89
9       54
10      14
11       4
27       1
19       1
16       1
43       1
14       1
12       1
13       1
18       1
Name: bhk, dtype: int64

In [27]:
df3['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [28]:
df3['total_sqft'].value_counts()

1200    843
1100    221
1500    204
2400    195
600     180
       ... 
5985      1
3580      1
2461      1
1437      1
4689      1
Name: total_sqft, Length: 2067, dtype: int64

<h4>The <u>try block</u> lets you test a block of code for errors.</h4>

<h4>The <u>except block</u> lets you handle the error.</h4>

In [29]:
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True

In [30]:
is_float(2)

True

In [31]:
is_float("2100 - 2850")

False

In [32]:
is_float("hello")

False

In [33]:
is_float(34.3)

True

In [34]:
float(2)

2.0

In [35]:
float("2")

2.0

In [36]:
float(22.5)

22.5

In [37]:
float("2 hi")
# could not convert string to float: '2 hi'

ValueError: could not convert string to float: '2 hi'

In [ ]:
float(200 - 400)

Using the ~ means to filter the values

In this case we want to filter the total_sqft so that it will only show numbers that are float 

The <b>tilde (~)</b> operator is used for “not” logic in filtering. If we add the tilde operator before the filter expression

the rows that do not fit the condition are returned.

In [57]:
df3[~df3['total_sqft'].apply(is_float)].head(10)

location       size      total_sqft  bath    price  bhk
30            Yelahanka      4 BHK     2100 - 2850   4.0  186.000    4
122              Hebbal      4 BHK     3067 - 8156   4.0  477.000    4
137  8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005    2
165            Sarjapur      2 BHK     1145 - 1340   2.0   43.490    2
188            KR Puram      2 BHK     1015 - 1540   2.0   56.800    2
410             Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500    1
549         Hennur Road      2 BHK     1195 - 1440   2.0   63.770    2
648             Arekere  9 Bedroom       4125Perch   9.0  265.000    9
661           Yelahanka      2 BHK     1120 - 1145   2.0   48.130    2
672        Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000    4

In [46]:
df3.shape

(13246, 6)

In [45]:
df3.shape

(13246, 6)

Loc and iloc

The loc and iloc methods are used to select rows or columns based on index or label.

    loc: select rows or columns using labels
    iloc: select rows or columns using indices

In [40]:
# find a single location of an item, loc = location 
df3.loc[30]

location        Yelahanka
size                4 BHK
total_sqft    2100 - 2850
bath                  4.0
price               186.0
bhk                     4
Name: 30, dtype: object

In [59]:
df3.iloc[3:5, :]  #rows 3 and 4, all columns

location   size total_sqft  bath  price  bhk
3  Lingadheeranahalli  3 BHK       1521   3.0   95.0    3
4            Kothanur  2 BHK       1200   2.0   51.0    2

In [61]:
df3.loc[3:5, :] #rows 3 and 4, all columns

location   size total_sqft  bath  price  bhk
3  Lingadheeranahalli  3 BHK       1521   3.0   95.0    3
4            Kothanur  2 BHK       1200   2.0   51.0    2
5          Whitefield  2 BHK       1170   2.0   38.0    2

In [41]:
# of df3, give me the indexes where df3.bhk is greather than 20  
df3[df3.bhk>20]

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

In [44]:
df3.shape

(13246, 6)

In [48]:
# you can set dataframe like a json/javascript object /library 
dfx = pd.DataFrame({
'name':['Jane','John','Ashley','Mike','Emily','Jack','Catlin'],
'ctg':['A','A','C','B','B','C','B'],
'val':np.random.random(7).round(2),
'val2':np.random.randint(1,10, size=7)
})

In [49]:
dfx

name ctg   val  val2
0    Jane   A  0.69     3
1    John   A  0.32     3
2  Ashley   C  0.01     4
3    Mike   B  0.86     3
4   Emily   B  0.93     1
5    Jack   C  0.65     9
6  Catlin   B  0.61     6

In [51]:
# Logical operators
# We can use the logical operators on column values to filter rows.
dfx[dfx.val > 0.5]

name ctg   val  val2
0    Jane   A  0.69     3
3    Mike   B  0.86     3
4   Emily   B  0.93     1
5    Jack   C  0.65     9
6  Catlin   B  0.61     6

In [54]:
# Only the names that come after ‘Jane’ in alphabetical order are selecte
dfx[dfx.name > 'Jane']

name ctg   val  val2
1  John   A  0.32     3
3  Mike   B  0.86     3

In [55]:
# Multiple logical operators 
dfx[(dfx.val > 0.5) & (dfx.val2 == 1)]

name ctg   val  val2
4  Emily   B  0.93     1

In [56]:
# The “&” signs stands for “and” , the “|” stands for “or”.
dfx[(dfx.val < 0.5) | (dfx.val2 == 7)]

name ctg   val  val2
1    John   A  0.32     3
2  Ashley   C  0.01     4

In [62]:
df3[~df3['total_sqft'].apply(is_float)].head(10)

location       size      total_sqft  bath    price  bhk
30            Yelahanka      4 BHK     2100 - 2850   4.0  186.000    4
122              Hebbal      4 BHK     3067 - 8156   4.0  477.000    4
137  8th Phase JP Nagar      2 BHK     1042 - 1105   2.0   54.005    2
165            Sarjapur      2 BHK     1145 - 1340   2.0   43.490    2
188            KR Puram      2 BHK     1015 - 1540   2.0   56.800    2
410             Kengeri      1 BHK  34.46Sq. Meter   1.0   18.500    1
549         Hennur Road      2 BHK     1195 - 1440   2.0   63.770    2
648             Arekere  9 Bedroom       4125Perch   9.0  265.000    9
661           Yelahanka      2 BHK     1120 - 1145   2.0   48.130    2
672        Bettahalsoor  4 Bedroom     3090 - 5002   4.0  445.000    4

Above shows that total_sqft can be a range (e.g. 2100-2850). For such case we can just take average of min and max value in the range. There are other cases such as 34.46Sq. Meter which one can convert to square ft using unit conversion. I am going to just drop such corner cases to keep things simple

In [70]:
def convert_sqft_to_num(x):
    # split tokens, most is seperated by a dash (-)
    tokens = x.split('-')
    if len(tokens) == 2:
        # if two variables, add them
        return (float(tokens[0])+float(tokens[1]))/2
    try:
        # if single variable, return variable 
        return float(x)
        # if neither, return none 
    except:
        return None   

.copy() allows you to copy a dataset, duplicate but do not connect 

a copy is sometimes needed so one can change one copy without changing the other

    copy.copy(x)
        Return a shallow copy of x.
        
    copy.deepcopy(x[, memo])
        Return a deep copy of x.
        
more info on copy: https://docs.python.org/3/library/copy.html

In [69]:
df4 = df3.copy()
# create a new column total_sqft 
df4.total_sqft = df4.total_sqft.apply(convert_sqft_to_num)
# can also be written 
#df4['total_sqft'] = df4['total_sqft'].apply(convert_sqft_to_num)

# new dataframe equals df4 w/o null values
# notnull() -> checks if value is null or not, return true/false 
df4 = df4[df4.total_sqft.notnull()]
df4.head(2)

location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4